cmd/snap-repair/repair.go

```diff
@@ -0,0 +1,174 @@
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/asserts/sysdb"
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/release"
+)
+
+var (
+	execCommand = exec.Command
+)
+
+type repair struct {
+	ra *asserts.Repair
+}
+
+func (r *repair) dir() string {
+	return filepath.Join(dirs.SnapRepairDir, r.ra.RepairID())
+}
+
+func (r *repair) ranStamp() string {
+	return filepath.Join(r.dir(), "ran")
+}
+
+func (r *repair) script() string {
+	return filepath.Join(r.dir(), "script")
+}
+
+func (r *repair) log() string {
+	return filepath.Join(r.dir(), fmt.Sprintf("%s.output", r.ra.RepairID()))
+}
+
+func (r *repair) wasRun() bool {
+	return osutil.FileExists(r.ranStamp())
+}
+
+func (r *repair) run() error {
+	// FIXME: deal with disk-full conditions in some way?
+	if err := os.MkdirAll(r.dir(), 0755); err != nil {
+		return err
+	}
+
+	logger.Noticef("starting repair %s", r.ra.RepairID())
+	if err := ioutil.WriteFile(r.script(), r.ra.Body(), 0755); err != nil {
+		return err
+	}
+	output, err := execCommand(r.script()).CombinedOutput()
+	if err := ioutil.WriteFile(r.log(), output, 0644); err != nil {
+		logger.Noticef("cannot write log: %s", err)
+	}
+	if err := ioutil.WriteFile(r.ranStamp(), nil, 0644); err != nil {
+		logger.Noticef("cannot write stamp: %s", err)
+	}
+	logger.Noticef("finished repair %s, logs in ", r.ra.RepairID(), r.log())
+
+	return err
+}
+
+type byRepairID []*asserts.Repair
+
+func (a byRepairID) Len() int {
+	return len(a)
+}
+func (a byRepairID) Swap(i, j int) {
+	a[i], a[j] = a[j], a[i]
+}
+
+// FIXME: move into the assertion itself?
+func splitID(id string) (string, int) {
+	l := strings.SplitN(id, "-", 2)
+	prefix := l[0]
+	seq, err := strconv.Atoi(l[1])
+	if err != nil {
+		panic(err)
+	}
+	return prefix, seq
+}
+
+func (a byRepairID) Less(i, j int) bool {
+	aPrefix, aSeq := splitID(a[i].RepairID())
+	bPrefix, bSeq := splitID(a[j].RepairID())
+	if aPrefix == bPrefix {
+		return aSeq < bSeq
+	}
+	return aPrefix < bPrefix
+}
+
+// FIXME: bypass the assertion DB entirely and collect all repair
+//        bits in /var/lib/snapd/repair/
+// FIXME: create a copy of the critical assertion code to protect
```

Comment:
I think that we should treat repair code change as a bootloader update. We should not do it unless we actually want to. It should not be automatically done whenever we change snapd itself. It should be extensively tested automatically as well as manually and only updated if we found an issue that needs addressing. The risk of just rebuilding this from the varying code all the time is that, as our last line of defence, it will fail us in  the only moment where it needs to work.